File: core/src/main/java/me/devsnox/planetsystem/core/planet/BaseLoadedPlanet.java
package me.devsnox.planetsystem.core.planet;

import com.boydti.fawe.FaweAPI;
import com.boydti.fawe.object.schematic.Schematic;
import com.sk89q.worldedit.regions.CuboidRegion;
import lombok.Data;
import lombok.EqualsAndHashCode;
import me.devsnox.planetsystem.api.holder.Holder;
import me.devsnox.planetsystem.api.location.PlanetLocation;
import me.devsnox.planetsystem.api.location.Region;
import me.devsnox.planetsystem.api.planet.LoadedPlanet;
import me.devsnox.planetsystem.api.planet.Planet;
import me.devsnox.planetsystem.core.location.BaseRegion;
import org.bukkit.Location;
import org.bukkit.util.Vector;

import java.util.UUID;
import java.util.function.Consumer;

@EqualsAndHashCode(callSuper = true)
@Data
public class BaseLoadedPlanet extends BasePlanet implements LoadedPlanet {

    private final Planet planet;
    private final Region inner;
    private final Region outer;
    private final Location middle;

    public BaseLoadedPlanet(Planet planet, Location middle, int maxSize) {
        super(planet.getUniqueID(), planet.getName(), planet.getOwnerUniqueID(), planet.getMembers(), planet.getSize(), planet.getSpawnLocation());
        this.planet = planet;
        this.middle = middle;

        byte size = planet.getSize();
        org.bukkit.util.Vector vSize = new org.bukkit.util.Vector(size, size, size);
        org.bukkit.util.Vector vMaxSize = new org.bukkit.util.Vector(maxSize, maxSize, maxSize);

        PlanetLocation innerMin = PlanetLocation.create(vSize.clone().multiply(-1)/*.add(new Vector(0, 1, 0))*/, middle.getYaw(), middle.getPitch(), planet.getUniqueID());
        PlanetLocation innerMax = PlanetLocation.create(vSize.clone().subtract(new Vector(1, 1, 1)), middle.getYaw(), middle.getPitch(), planet.getUniqueID());

        PlanetLocation outerMin = PlanetLocation.create(vMaxSize.clone().multiply(-1), middle.getYaw(), middle.getPitch(), planet.getUniqueID());
        PlanetLocation outerMax = PlanetLocation.create(vMaxSize, middle.getYaw(), middle.getPitch(), planet.getUniqueID());

        this.inner = new BaseRegion(innerMin, innerMax);
        this.outer = new BaseRegion(outerMin, outerMax);
    }

    @Override
    public Location getMiddle() {
        return this.middle.clone();
    }

    @Override
    public Schematic getSchematic() {
        Schematic schematic = new Schematic(new CuboidRegion(FaweAPI.getWorld(Holder.Impl.holder.getWorld().getName()), this.inner.getMin().toWEVector(), this.inner.getMax().toWEVector()));
        System.out.println(schematic);
        System.out.println(schematic.getClipboard());
        System.out.println(schematic.getClipboard().getDimensions());
        System.out.println(schematic.getClipboard().getOrigin());
        System.out.println(schematic.getClipboard().getRegion());
        System.out.println(schematic.getClipboard().getEntities());

        //schematic.getClipboard().setOrigin(new com.sk89q.worldedit.Vector(this.middle.getX(), this.middle.getY(), this.middle.getZ()));

        return schematic;
    }

    @Override
    public UUID getUniqueID() {
        return this.planet.getUniqueID();
    }

    @Override
    public UUID getOwnerUniqueID() {
        return this.planet.getOwnerUniqueID();
    }

    @Override
    public String getName() {
        return this.planet.getName();
    }

    @Override
    public PlanetLocation getSpawnLocation() {
        return this.planet.getSpawnLocation();
    }

    @Override
    public void setSpawnLocation(PlanetLocation planetLocation) {
        this.planet.setSpawnLocation(planetLocation);
    }

    @Override
    public void load(Consumer<LoadedPlanet> result) {
        result.accept(this);
    }

}

File: core/src/main/kotlin/me/devsnox/planetsystem/core/planet/BaseLoadedPlanet.kt
package me.devsnox.planetsystem.core.planet

import com.boydti.fawe.`object`.schematic.Schematic
import com.sk89q.worldedit.regions.CuboidRegion
import lombok.Data
import lombok.EqualsAndHashCode
import me.devsnox.planetsystem.api.functions.toWEVector
import me.devsnox.planetsystem.api.functions.toWEWorld
import me.devsnox.planetsystem.api.holder.Holder.Impl.holder
import me.devsnox.planetsystem.api.location.PlanetLocation
import me.devsnox.planetsystem.api.location.Region
import me.devsnox.planetsystem.api.planet.LoadedPlanet
import me.devsnox.planetsystem.api.planet.Planet
import me.devsnox.planetsystem.core.location.BaseRegion
import org.bukkit.Location
import org.bukkit.util.Vector

@EqualsAndHashCode(callSuper = true)
@Data
class BaseLoadedPlanet(
		planet: Planet,
		middle: Location,
		maxSize: Int
) : LoadedPlanet, BasePlanet(planet.uniqueID, planet.name, planet.ownerUniqueID, planet.members, planet.size, planet.spawnLocation) {

	override val middle = middle
		get() = field.clone()
	override val inner: Region
	override val outer: Region

	override val schematic: Schematic
		get() = Schematic(CuboidRegion(holder.world.toWEWorld(), inner.min.toWEVector(), inner.max.toWEVector())).apply {
			clipboard!!.origin = middle.toWEVector()
		}

	/**
	 * @author Lars Artmann | LartyHD
	 * Created by Lars Artmann | LartyHD on 12.02.2019 17:55.
	 * Current Version: 1.0 (12.02.2019 - 12.02.2019)
	 */
	override var size: Byte
		get() = super.size
		set(size) {
			super.size = size
			val vector = Vector(size.toInt(), size - 1, size.toInt())
			inner.max.vector = vector.subtract(Vector(1, 0, 1))
			inner.min.vector = vector.clone().multiply(-1)
		}

	init {

		val innerMin = size.toBukkitVector().multiply(-1).toPlanetLocation()
		val innerMax = size.toBukkitVector().subtract(Vector(1, 1, 1)).toPlanetLocation()

		val outerMin = maxSize.toBukkitVector().multiply(-1).toPlanetLocation()
		val outerMax = maxSize.toBukkitVector().toPlanetLocation()

		inner = BaseRegion(innerMin, innerMax)
		outer = BaseRegion(outerMin, outerMax)

	}

	override fun load(result: (LoadedPlanet) -> Unit) = super<LoadedPlanet>.load(result)

	//TODO Add to Darkness
	@Suppress("MemberVisibilityCanBePrivate")
	fun Number.toBukkitVector() = toDouble().run { Vector(this, this, this) }

	private fun Vector.toPlanetLocation() = PlanetLocation(uniqueID, this, middle.yaw, middle.pitch)

}
